# 1. 데이터 정보 확인

In [2]:
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt

PATH='C:/Users/castl/Desktop/2020_job_academy/project/lg_sys/'
f_err=pd.read_csv(PATH+'train_err_data.csv')
f_prob=pd.read_csv(PATH+'train_problem_data.csv')
f_qual=pd.read_csv(PATH+'train_quality_data.csv')

C:\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3146: DtypeWarning: Columns (10,11,12) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
f_err.head(5)

user_id            time model_nm       fwver  errtype errcode
0    10000  20201101025616  model_3  05.15.2138       15       1
1    10000  20201101030309  model_3  05.15.2138       12       1
2    10000  20201101030309  model_3  05.15.2138       11       1
3    10000  20201101050514  model_3  05.15.2138       16       1
4    10000  20201101050515  model_3  05.15.2138        4       0

In [4]:
f_prob.head(5)

user_id            time
0    19224  20201102200000
1    23664  20201116140000
2    15166  20201114130000
3    12590  20201108210000
4    15932  20201103210000

In [5]:
f_qual.head(5)

time  user_id       fwver  quality_0  quality_1  quality_2  \
0  20201129090000    10000  05.15.2138        0.0          0        0.0   
1  20201129090000    10000  05.15.2138        0.0          0        0.0   
2  20201129090000    10000  05.15.2138        0.0          0        0.0   
3  20201129090000    10000  05.15.2138        0.0          0        0.0   
4  20201129090000    10000  05.15.2138        0.0          0        0.0   

   quality_3  quality_4 quality_5  quality_6 quality_7 quality_8 quality_9  \
0          0          0         0          0         0         0         0   
1          0          0         0          0         0         0         0   
2          0          0         0          0         0         0         0   
3          0          0         0          0         0         0         0   
4          0          0         0          0         0         0         0   

  quality_10  quality_11  quality_12  
0          4           0           0  
1          4           0           0  
2          4           0           0  
3          4           0           0  
4          4           0           0

In [6]:
f_err.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16554663 entries, 0 to 16554662
Data columns (total 6 columns):
 #   Column    Dtype 
---  ------    ----- 
 0   user_id   int64 
 1   time      int64 
 2   model_nm  object
 3   fwver     object
 4   errtype   int64 
 5   errcode   object
dtypes: int64(3), object(3)
memory usage: 757.8+ MB


In [7]:
# null값 확인
# errcode가 null값이 있다는 것 확인

f_err.isna().any()

user_id     False
time        False
model_nm    False
fwver       False
errtype     False
errcode      True
dtype: bool

In [8]:
# null값은 1개 인것 확인, 전처리에서 null값 처리 예정
f_err['errcode'].isna().sum()

1

In [9]:
# 각 feature마다 unique값 및 갯수 확인
for i in f_err.columns:
    print(i,'\n')
    print(f_err[i].unique(),'\n',len(f_err[i].unique()))

user_id 

[10000 10001 10002 ... 24997 24998 24999] 
 15000
time 

[20201101025616 20201101030309 20201101050514 ... 20201118123022
 20201129070715 20201129070741] 
 2566163
model_nm 

['model_3' 'model_2' 'model_0' 'model_1' 'model_7' 'model_4' 'model_5'
 'model_8' 'model_6'] 
 9
fwver 

['05.15.2138' '04.33.1185' '04.33.1261' '04.22.1750' '04.22.1778'
 '04.16.3553' '04.33.1149' '04.16.3571' '05.66.3237' '05.66.3571'
 '03.11.1149' '04.22.1684' '03.11.1167' '04.82.1684' '04.82.1778'
 '04.33.1171' '04.73.2237' '10' '04.82.1730' '04.73.2571' '8.5.3'
 '04.22.1666' '03.11.1141' '05.15.2120' '04.33.1125' '04.16.3439'
 '04.22.1442' '04.33.1095' '04.16.3569' '05.15.2090' '05.15.3104'
 '05.15.2122' '04.22.1656' '04.16.2641' '05.15.2114' '04.16.3345'
 '05.15.2092'] 
 37
errtype 

[15 12 11 16  4 26 10 14  3 13  6  7 27 28 31 33 22 41  5 40 34 37 36 23
 18 20 19 21 42 24 25 17 38 39 35 32 30  2  1  9  8] 
 41
errcode 

['1' '0' '2' ... '4526' '3965' '25999'] 
 2806


### f_err 파일

- user_id는 10000번부터 24999까지 15000개인것 확인
- time은 년/월/시/분/초 로 구성되어있어 datetime 전처리 예정
- model은 0번부터 8번까지 9가지 존재
- fwver 37개 존재
- errtype은 1번부터 42번까지 29번을 제외한 41개 존재
- errcode는 2806개 존재
- errcode에서 null값 1개 확인, null값 수정 혹은 삭제 예정

In [10]:
f_prob.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5429 entries, 0 to 5428
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype
---  ------   --------------  -----
 0   user_id  5429 non-null   int64
 1   time     5429 non-null   int64
dtypes: int64(2)
memory usage: 85.0 KB


In [11]:
f_prob.isna().any()

user_id    False
time       False
dtype: bool

In [12]:
for i in f_prob.columns:
    print(i,'\n')
    print(f_prob[i].unique(),'\n',len(f_prob[i].unique()))

user_id 

[19224 23664 15166 ... 19114 21505 18822] 
 5000
time 

[20201102200000 20201116140000 20201114130000 20201108210000
 20201103210000 20201119150000 20201121110000 20201111160000
 20201118120000 20201127230000 20201117120000 20201129110000
 20201119180000 20201112150000 20201116150000 20201130160000
 20201130220000 20201118220000 20201104160000 20201118200000
 20201130120000 20201127170000 20201121220000 20201102110000
 20201102130000 20201113190000 20201102170000 20201120190000
 20201114010000 20201124160000 20201109110000 20201107120000
 20201116160000 20201124150000 20201102120000 20201127120000
 20201123150000 20201104130000 20201126200000 20201106180000
 20201109210000 20201126150000 20201127150000 20201121170000
 20201124120000 20201121150000 20201116200000 20201109150000
 20201109140000 20201103180000 20201126110000 20201107230000
 20201110170000 20201123120000 20201119220000 20201110110000
 20201117130000 20201109200000 20201116120000 20201113230000
 20201107180000 202

### f_prob파일

- null값 없음
- 불만 접수된 서로 다른 user_id는 5000개
- 서로다른 시간대는 년/월/시/분/초 로 datetime 전처리 예정, 총 488개

In [13]:
f_qual.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 828624 entries, 0 to 828623
Data columns (total 16 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   time        828624 non-null  int64  
 1   user_id     828624 non-null  int64  
 2   fwver       788544 non-null  object 
 3   quality_0   684192 non-null  float64
 4   quality_1   828624 non-null  int64  
 5   quality_2   788511 non-null  float64
 6   quality_3   828624 non-null  int64  
 7   quality_4   828624 non-null  int64  
 8   quality_5   828604 non-null  object 
 9   quality_6   828624 non-null  int64  
 10  quality_7   828624 non-null  object 
 11  quality_8   828624 non-null  object 
 12  quality_9   828624 non-null  object 
 13  quality_10  828624 non-null  object 
 14  quality_11  828624 non-null  int64  
 15  quality_12  828624 non-null  int64  
dtypes: float64(2), int64(8), object(6)
memory usage: 101.2+ MB


In [14]:
#null값 확인
f_qual.isna().any()

time          False
user_id       False
fwver          True
quality_0      True
quality_1     False
quality_2      True
quality_3     False
quality_4     False
quality_5      True
quality_6     False
quality_7     False
quality_8     False
quality_9     False
quality_10    False
quality_11    False
quality_12    False
dtype: bool

In [15]:
for i in f_qual.columns:
    if f_qual[i].isna().any()==True:
        print(i,f_qual[i].isna().sum())

fwver 40080
quality_0 144432
quality_2 40113
quality_5 20


In [16]:
for i in f_qual.columns:
    print(i,'\n')
    print(f_qual[i].unique(),'\n',len(f_qual[i].unique()))

time 

[20201129090000 20201130210000 20201104110000 ... 20201126034000
 20201113065000 20201116061000] 
 4319
user_id 

[10000 10002 10004 ... 24993 24995 24997] 
 8281
fwver 

['05.15.2138' '04.22.1750' '04.16.3553' '04.33.1261' '04.22.1778'
 '04.33.1185' '04.16.3571' '05.66.3571' '03.11.1149' '03.11.1167'
 '04.82.1684' '04.82.1778' '04.33.1149' nan '05.66.3237' '04.73.2237'
 '09.17.1431' '04.22.1684' '05.15.2120' '04.33.1125' '05.15.2122'
 '04.22.1666' '04.22.1656' '04.16.3439' '04.73.2571' '05.15.2114'
 '04.16.3345' '04.22.1442'] 
 28
quality_0 

[ 0.00000e+00  2.00000e+00 -1.00000e+00          nan  1.00000e+00
  3.00000e+00  7.00000e+00  6.00000e+00  4.08800e+03  1.29000e+02
  3.18600e+03  2.96000e+02  1.70000e+01  6.30000e+01  1.00000e+01
  9.04000e+02  4.00000e+00  3.79000e+03  1.62180e+04  1.06280e+04
  1.04000e+02  7.80000e+01  3.79000e+02  3.46000e+02  2.15000e+02
  2.61000e+02  3.34000e+02  2.50000e+02  1.60000e+01  3.10000e+01
  8.70000e+01  2.98000e+02  1.50000e+01  7.7000

[0] 
 1
quality_5 

['0' '4' '8' ... '98,175' '7,998' '9,558'] 
 4745
quality_6 

[  0   1  -1  44  87  27   9   4 348   2   5  68 204  11  49  62   8  67
  19 139 196  82  37   6 600 398  24  12  46 110  42  36 599 590 485  41
  33  30 283 317  47 105  25  75   7  13  54  92  64  95  72 333  70 519
 161 132 301 109  22  89  28 264 158 120  51 117   3  21  15  10  98 178
  23  18  35 153 156 367 119  88 202 305 175  29  14  20 218  17  77  32
  81  26  31  57  60  50  66 215 199 113 152  74  63  45 172 318 269 182
  96 540 564 198  71 254 304 136  61 523 167 381 224  40 566 275 598 277
  52 436  16 130 459  38  55  53 123 289 525 395 190 162 281 467 524 550
 539 251  99 282 447 267 347 440 571 479 423 184 480 276  91  80 101  43
 499 500 169 179 195 201 188  97 260  48  69  34  79  39 102 163 255 144
 103  56 320 428 559  94 185 191 129 486 285 585  86 171 391 107  85 207
  93 127 150  84 206 567 104 241 145 125  78  59 538 306 380 326 115 128
 116 226  65 106 338 429 220 232 580 108 4

[ 0 -1  1  3  2  4  5  6  9  8 14  7] 
 12
quality_12 

[ 0  1  2  3  5  7  6  8  4 10 12 13 14 11] 
 14


### f_qual파일

- fwver,quality0,2,5에서 null값 다수 발견
- 서로다른 user_id는 8281개이다.

# 2. 전처리(Data Post Processing

In [17]:
# 시간 datetime tyoe으로 변환

def change_time(f):
    l=[]
    for i in range(len(f)):
        t=str(f['time'][i])
        tmp=t[0:4]+'-'+t[4:6]+'-'+t[6:8]+' '+t[8:10]+':'+t[10:12]+':'+t[12:14]
        l.append(tmp)
        
    f['time']=l
    f['time']=pd.to_datetime(f['time'],format='%Y-%m-%d %H:%M:%S')
    
    return f

In [18]:
change_time(f_err)

user_id                time model_nm       fwver  errtype errcode
0           10000 2020-11-01 02:56:16  model_3  05.15.2138       15       1
1           10000 2020-11-01 03:03:09  model_3  05.15.2138       12       1
2           10000 2020-11-01 03:03:09  model_3  05.15.2138       11       1
3           10000 2020-11-01 05:05:14  model_3  05.15.2138       16       1
4           10000 2020-11-01 05:05:15  model_3  05.15.2138        4       0
...           ...                 ...      ...         ...      ...     ...
16554658    24999 2020-11-30 16:30:51  model_3  05.15.2138       15       1
16554659    24999 2020-11-30 17:26:25  model_3  05.15.2138       16       1
16554660    24999 2020-11-30 17:26:25  model_3  05.15.2138        4       0
16554661    24999 2020-11-30 17:26:31  model_3  05.15.2138        4       0
16554662    24999 2020-11-30 21:06:25  model_3  05.15.2138       15       1

[16554663 rows x 6 columns]

In [19]:
change_time(f_prob)

user_id                time
0       19224 2020-11-02 20:00:00
1       23664 2020-11-16 14:00:00
2       15166 2020-11-14 13:00:00
3       12590 2020-11-08 21:00:00
4       15932 2020-11-03 21:00:00
...       ...                 ...
5424    20167 2020-11-25 12:00:00
5425    16270 2020-11-10 12:00:00
5426    19114 2020-11-06 23:00:00
5427    21505 2020-11-04 11:00:00
5428    18822 2020-11-02 12:00:00

[5429 rows x 2 columns]

In [20]:
change_time(f_qual)

time  user_id       fwver  quality_0  quality_1  \
0      2020-11-29 09:00:00    10000  05.15.2138        0.0          0   
1      2020-11-29 09:00:00    10000  05.15.2138        0.0          0   
2      2020-11-29 09:00:00    10000  05.15.2138        0.0          0   
3      2020-11-29 09:00:00    10000  05.15.2138        0.0          0   
4      2020-11-29 09:00:00    10000  05.15.2138        0.0          0   
...                    ...      ...         ...        ...        ...   
828619 2020-11-24 03:30:00    24997  04.22.1778        0.0          0   
828620 2020-11-24 03:30:00    24997  04.22.1778        0.0          0   
828621 2020-11-24 03:30:00    24997  04.22.1778        0.0          0   
828622 2020-11-24 03:30:00    24997  04.22.1778        0.0          0   
828623 2020-11-24 03:30:00    24997  04.22.1778        0.0          0   

        quality_2  quality_3  quality_4 quality_5  quality_6 quality_7  \
0             0.0          0          0         0          0         0   
1             0.0          0          0         0          0         0   
2             0.0          0          0         0          0         0   
3             0.0          0          0         0          0         0   
4             0.0          0          0         0          0         0   
...           ...        ...        ...       ...        ...       ...   
828619        0.0          0          0         1          0         0   
828620        0.0          0          0         0          0         0   
828621        0.0          0          0         3          0         0   
828622        0.0          0          0         0          0         0   
828623        0.0          0          0         9          0         0   

       quality_8 quality_9 quality_10  quality_11  quality_12  
0              0         0          4           0           0  
1              0         0          4           0           0  
2              0         0          4           0           0  
3              0         0          4           0           0  
4              0         0          4           0           0  
...          ...       ...        ...         ...         ...  
828619         0         0         17           0           0  
828620         0         0         17           0           0  
828621         0         0         17           0           0  
828622         0         0         17           0           0  
828623         0         0         17           0           0  

[828624 rows x 16 columns]

In [21]:
f_err.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16554663 entries, 0 to 16554662
Data columns (total 6 columns):
 #   Column    Dtype         
---  ------    -----         
 0   user_id   int64         
 1   time      datetime64[ns]
 2   model_nm  object        
 3   fwver     object        
 4   errtype   int64         
 5   errcode   object        
dtypes: datetime64[ns](1), int64(2), object(3)
memory usage: 757.8+ MB


In [22]:
f_prob.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5429 entries, 0 to 5428
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype         
---  ------   --------------  -----         
 0   user_id  5429 non-null   int64         
 1   time     5429 non-null   datetime64[ns]
dtypes: datetime64[ns](1), int64(1)
memory usage: 85.0 KB


In [23]:
f_qual.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 828624 entries, 0 to 828623
Data columns (total 16 columns):
 #   Column      Non-Null Count   Dtype         
---  ------      --------------   -----         
 0   time        828624 non-null  datetime64[ns]
 1   user_id     828624 non-null  int64         
 2   fwver       788544 non-null  object        
 3   quality_0   684192 non-null  float64       
 4   quality_1   828624 non-null  int64         
 5   quality_2   788511 non-null  float64       
 6   quality_3   828624 non-null  int64         
 7   quality_4   828624 non-null  int64         
 8   quality_5   828604 non-null  object        
 9   quality_6   828624 non-null  int64         
 10  quality_7   828624 non-null  object        
 11  quality_8   828624 non-null  object        
 12  quality_9   828624 non-null  object        
 13  quality_10  828624 non-null  object        
 14  quality_11  828624 non-null  int64         
 15  quality_12  828624 non-null  int64         
dtypes:

- f_err,f_prob, f_qual의 time을 datetime 형태로 변환

In [24]:
# 결측치가 있는 행 찾기
f_err[f_err['errcode'].isna()==True]

user_id                time model_nm       fwver  errtype errcode
3825744    13639 2020-11-21 19:17:18  model_2  04.33.1261        5     NaN

In [25]:
# 해당 user_id가 사용자 불만에 신고된 적이있는지 확인(중요도 판단 요소)
f_prob[f_prob['user_id']==13639]

user_id                time
3988    13639 2020-11-17 21:00:00

In [26]:
# errtype에 따라 errcode가 같은지 확인,
# 확인 결과 상관없음
f_err[f_err['errtype']==5]

user_id                time model_nm       fwver  errtype  errcode
330         10001 2020-11-02 00:23:05  model_2  04.33.1185        5  B-A8002
331         10001 2020-11-02 00:23:10  model_2  04.33.1185        5  B-A8002
332         10001 2020-11-02 00:23:10  model_2  04.33.1185        5  B-A8002
333         10001 2020-11-02 00:23:11  model_2  04.33.1185        5  B-A8002
334         10001 2020-11-02 00:23:11  model_2  04.33.1185        5  B-A8002
...           ...                 ...      ...         ...      ...      ...
16554293    24999 2020-11-11 03:04:46  model_3  05.15.2138        5  Q-64002
16554332    24999 2020-11-13 03:04:52  model_3  05.15.2138        5  Q-64002
16554406    24999 2020-11-17 03:04:58  model_3  05.15.2138        5  S-65002
16554448    24999 2020-11-19 03:04:48  model_3  05.15.2138        5  Q-64002
16554634    24999 2020-11-29 03:04:53  model_3  05.15.2138        5  Q-64002

[952920 rows x 6 columns]

In [27]:
# user_id가 동일한 errcode를 보이는지 확인,
# 확인 결과 errtype, errcode가 다양함(규칙성 없음)
x=f_err[f_err['user_id']==13639]
x

user_id                time model_nm       fwver  errtype errcode
3823076    13639 2020-11-01 00:00:04  model_2  04.33.1185        2       0
3823077    13639 2020-11-01 00:05:33  model_2  04.33.1185        2       0
3823078    13639 2020-11-01 00:29:14  model_2  04.33.1185        2       1
3823079    13639 2020-11-01 00:29:24  model_2  04.33.1185        2       0
3823080    13639 2020-11-01 00:31:21  model_2  04.33.1185        2       1
...          ...                 ...      ...         ...      ...     ...
3825960    13639 2020-11-30 22:03:02  model_2  04.33.1261       16       1
3825961    13639 2020-11-30 22:03:03  model_2  04.33.1261       40       0
3825962    13639 2020-11-30 22:03:06  model_2  04.33.1261       31       1
3825963    13639 2020-11-30 22:03:18  model_2  04.33.1261       31       0
3825964    13639 2020-11-30 23:05:00  model_2  04.33.1261       15       1

[2889 rows x 6 columns]

In [28]:
# errcode 분포
x.groupby(['errcode'])['errcode'].count()

errcode
0          1245
1          1532
13            1
14            2
2            11
3            10
40013         1
78            1
79            3
8.0           2
80            4
B-A8002      61
H-51042       2
UNKNOWN       1
active        5
standby       7
Name: errcode, dtype: int64

In [29]:
f_err=f_err.dropna(axis=0,how='any')

In [30]:
f_err.isna().any()

user_id     False
time        False
model_nm    False
fwver       False
errtype     False
errcode     False
dtype: bool

- nan값이 있는 user_id 13639는 errcode 0,1중 하나로 추정 할 수 있지만 불확실 함으로 해당 행을 삭제한다.

In [31]:
f_qual=f_qual.dropna(subset=['quality_0','quality_2','quality_5'],how='any',axis=0)

In [32]:
f_qual.isna().any()

time          False
user_id       False
fwver         False
quality_0     False
quality_1     False
quality_2     False
quality_3     False
quality_4     False
quality_5     False
quality_6     False
quality_7     False
quality_8     False
quality_9     False
quality_10    False
quality_11    False
quality_12    False
dtype: bool

In [33]:
f_qual.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 684139 entries, 0 to 828623
Data columns (total 16 columns):
 #   Column      Non-Null Count   Dtype         
---  ------      --------------   -----         
 0   time        684139 non-null  datetime64[ns]
 1   user_id     684139 non-null  int64         
 2   fwver       684139 non-null  object        
 3   quality_0   684139 non-null  float64       
 4   quality_1   684139 non-null  int64         
 5   quality_2   684139 non-null  float64       
 6   quality_3   684139 non-null  int64         
 7   quality_4   684139 non-null  int64         
 8   quality_5   684139 non-null  object        
 9   quality_6   684139 non-null  int64         
 10  quality_7   684139 non-null  object        
 11  quality_8   684139 non-null  object        
 12  quality_9   684139 non-null  object        
 13  quality_10  684139 non-null  object        
 14  quality_11  684139 non-null  int64         
 15  quality_12  684139 non-null  int64         
dtypes:

- quality0,2,5에서 nan값을 갖고 있는 행 제거

In [34]:
f_qual.groupby(['user_id','fwver'])['user_id'].count()

user_id  fwver     
10000    05.15.2138    24
10002    05.15.2138    96
10004    04.22.1750    24
10005    04.22.1750    24
10006    04.16.3553    12
                       ..
24992    04.22.1750    36
24993    05.15.2138    24
24995    04.33.1261    24
24997    04.22.1750    12
         04.22.1778    12
Name: user_id, Length: 9569, dtype: int64

In [35]:
f_err.groupby(['user_id','fwver'])['user_id'].count()

user_id  fwver     
10000    05.15.2138     317
10001    04.33.1185     379
         04.33.1261    1986
10002    05.15.2138     306
10003    04.33.1185      81
                       ... 
24997    04.22.1750     465
         04.22.1778     361
24998    04.22.1750       8
         04.22.1778     147
24999    05.15.2138     570
Name: user_id, Length: 24062, dtype: int64

- f_err과 f_qual의 fwver의 종류는 같지만, 갯수가 다르기 때문에 단일 fwver을 갖지 않는 user_id에 대해서는 nan값을 대채할 수 없다.

In [36]:
#각 user_id의 fwver 갯수 확인 

counting=[]
l=[]
for i in range(1,len(f_err)):
    if f_err['user_id'].values[i]==f_err['user_id'].values[i-1]:
        if f_err['fwver'].values[i] not in l:
            l.append(f_err['fwver'].values[i])
    else:
        counting.append(len(l))
        l=[]

In [37]:
len(counting)

14999

In [38]:
# f_qual의 fwver이 counting이 1개인것 만 nan값 대신 fwver 값 변경

for i in range(len(f_qual)):
    if f_qual['fwver'].isna().values[i]:
        if counting[f_qual['user_id'].values[i]-10000]==1:
            a=f_qual['user_id'].values[i]
            f_qual['fwver'][i]=f_err[f_err["user_id"]==a]['fwver'].values[0]
            

In [39]:
# nan값 없어졌는지 확인

f_qual.isna().any()

time          False
user_id       False
fwver         False
quality_0     False
quality_1     False
quality_2     False
quality_3     False
quality_4     False
quality_5     False
quality_6     False
quality_7     False
quality_8     False
quality_9     False
quality_10    False
quality_11    False
quality_12    False
dtype: bool

- f_qual의 fwver은 각 user_id별로 1개의 fwver갖고 있는 것만 nan값을 해당 fwver로 변환
- 다중 fwver을 갖고 있는 user_id는 불확실하기에 fwver nan값이 있는 해당 행을 제거

# 3. 모델링전 전처리

### Target 생성

In [40]:
# total_user_id는 서로 다른 user_id갯수

total_user_id=len(f_err['user_id'].unique())
target=[0]*total_user_id

for i in f_prob.user_id.values:
    if target[i-10000]==0:
        target[i-10000]=1


In [139]:
len(target)

15000

In [142]:
target

[0,
 1,
 0,
 0,
 1,
 1,
 1,
 0,
 1,
 1,
 0,
 1,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 1,
 0,
 1,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 1,
 0,
 0,
 1,
 1,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 1,
 1,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 1,
 1,
 0,
 1,
 1,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 1,
 1,
 0,
 1,
 0,
 1,
 0,
 1,
 1,
 1,
 1,
 0,
 1,
 0,
 1,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 1,
 1,
 1,
 0,
 0,
 1,
 1,
 0,
 1,
 0,
 1,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 1,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 1,
 0,
 0,
 1,
 1,
 1,
 1,
 1,
 0,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 0,
 0,
 0,
 1,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 1,


In [150]:
# np로 변형후 reshape

r_target=np.array(target).reshape(len(target),-1)

r_target

array([[0],
       [1],
       [0],
       ...,
       [1],
       [1],
       [0]])

- 문제 신고가 됬으면 target에 1로 표시 없으면 0으로 표시함
- target은 list0부터14999까지가 user_id 10000부터 24999까지임
- reshape하여 target변수로써 기능을 할 수 있도록 재구성

### 상관관계 탐색

In [121]:
tmp_err=f_err.copy()

In [126]:
ALL_TARGET=[]

for i in f_err.user_id.values:
    if i in f_prob.user_id.unique():
        ALL_TARGET.append(1)
    else:
        ALL_TARGET.append(0)

In [127]:
tmp_err['TARGET']=ALL_TARGET

In [133]:
tmp_err

user_id                time model_nm       fwver  errtype errcode  \
0           10000 2020-11-01 02:56:16  model_3  05.15.2138       15       1   
1           10000 2020-11-01 03:03:09  model_3  05.15.2138       12       1   
2           10000 2020-11-01 03:03:09  model_3  05.15.2138       11       1   
3           10000 2020-11-01 05:05:14  model_3  05.15.2138       16       1   
4           10000 2020-11-01 05:05:15  model_3  05.15.2138        4       0   
...           ...                 ...      ...         ...      ...     ...   
16554658    24999 2020-11-30 16:30:51  model_3  05.15.2138       15       1   
16554659    24999 2020-11-30 17:26:25  model_3  05.15.2138       16       1   
16554660    24999 2020-11-30 17:26:25  model_3  05.15.2138        4       0   
16554661    24999 2020-11-30 17:26:31  model_3  05.15.2138        4       0   
16554662    24999 2020-11-30 21:06:25  model_3  05.15.2138       15       1   

          TARGET  
0              0  
1              0  
2              0  
3              0  
4              0  
...          ...  
16554658       0  
16554659       0  
16554660       0  
16554661       0  
16554662       0  

[16554662 rows x 7 columns]

In [141]:
# pearson보다 spearman 계수가 해당 데이터(순서형자료)에 맞다고 생각함

tmp_err.corr(method='pearson').style.background_gradient(cmap='coolwarm')

### feature 재정의

In [41]:
user=[]

for i in f_err.user_id.values:
    if i not in user:
        user.append(i)

In [48]:
x=pd.DataFrame(tmp_err['errtype'])
x

errtype
0              15
1              12
2              11
3              16
4               4
...           ...
16554658       15
16554659       16
16554660        4
16554661        4
16554662       15

[16554662 rows x 1 columns]

In [76]:
# id에 f_prob의 id만 넣음
# t_time에 f_prob의 id에서 같을 경우 최초 시간 기준(시간이 더 짧은것)으로 생성 

id=[]
t_time=[]
for idx,t in f_prob[['user_id','time']].values:
    if idx not in id:
        id.append(idx)
        t_time.append(t)
    else:
        if t_time[id.index(idx)]>t:
            t_time[id.index(idx)]=t


In [81]:
col_list=['user_id','time']
col_data={'user_id':id,
         'time':t_time}

r_prob=pd.DataFrame(col_data,columns=col_list)
r_prob

user_id                time
0       19224 2020-11-02 20:00:00
1       23664 2020-11-16 14:00:00
2       15166 2020-11-14 13:00:00
3       12590 2020-11-08 21:00:00
4       15932 2020-11-03 21:00:00
...       ...                 ...
4995    20167 2020-11-25 12:00:00
4996    16270 2020-11-10 12:00:00
4997    19114 2020-11-06 23:00:00
4998    21505 2020-11-04 11:00:00
4999    18822 2020-11-02 12:00:00

[5000 rows x 2 columns]

In [136]:
errtype_feature=[]
tmp=[0]*41
        
for i in range(1,len(f_err)):
    if f_err.user_id.values[i] in id:
        if f_err['user_id'].values[i]==f_err['user_id'].values[i-1]:
            if f_err.time.values[i]<=t_time[id.index(f_err.user_id.values[i])]:
                if f_err.errtype.values[i]<29:
                    tmp[f_err.errtype.values[i]-1]+=1
                else:
                    tmp[f_err.errtype.values[i]-2]+=1
            else:
                tmp=tmp
        else:
            errtype_feature.append(tmp)
            tmp=[0]*41
            if f_err.errtype.values[i]<29:
                tmp[f_err.errtype.values[i]-1]+=1
            else:
                tmp[f_err.errtype.values[i]-2]+=1
    else:
        if tmp_err['user_id'].values[i]==tmp_err['user_id'].values[i-1]:
            if f_err.errtype.values[i]<29:
                tmp[f_err.errtype.values[i]-1]+=1
            else:
                tmp[f_err.errtype.values[i]-2]+=1
        else:
            errtype_feature.append(tmp)
            tmp=[0]*41
            if f_err.errtype.values[i]<29:
                tmp[f_err.errtype.values[i]-1]+=1
            else:
                tmp[f_err.errtype.values[i]-2]+=1
                
errtype_feature.append(tmp)

In [158]:
r_feature=np.array(errtype_feature)
r_feature

array([[ 0,  0,  8, ...,  0,  0,  0],
       [ 0,  0,  0, ..., 74, 28,  0],
       [ 0,  0,  2, ...,  0,  0,  0],
       ...,
       [ 0,  0,  0, ..., 36,  0,  0],
       [ 0,  0,  0, ...,  2,  0,  0],
       [ 0,  0,  4, ...,  0,  0,  0]])

# 4. 모델링

In [160]:
trainX=r_feature
trainY=r_target
print(trainX.shape)
print(trainY.shape)

(15000, 41)
(15000, 1)


In [162]:
# Support vector machine

from sklearn.svm import SVC

svm = SVC(kernel='rbf',C=10,gamma=0.1,random_state=0)

In [163]:
### 교차검증

from sklearn.model_selection import KFold
from sklearn.metrics import accuracy_score

kfold=KFold(n_splits=5)
n_iter=0
r_accuracy=[]

for train_index, test_index in kfold.split(r_feature):
    X_train, X_test=r_feature[train_index],r_feature[test_index]
    Y_train, Y_test=r_target[train_index],r_target[test_index]
    
    svm.fit(X_train,Y_train)
    pred=svm.predict(X_test)
    n_iter+=1
    
    accuracy=np.round(accuracy_score(Y_test,pred),4)
    train_size=X_train.shape[0]
    test_size=X_test.shape[0]
    print('{0} 교차검증 정확도 :{1}, 학습 데이터 크기:{2}, 검증 데이터 크기:{3}'.format(n_iter,accuracy,train_size,test_size))
    
    r_accuracy.append(accuracy)
    
print('\n 평균검증 정확도:',np.mean(r_accuracy))

C:\Anaconda3\lib\site-packages\sklearn\utils\validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


1 교차검증 정확도 :0.6883, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\sklearn\utils\validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


2 교차검증 정확도 :0.6803, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\sklearn\utils\validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


3 교차검증 정확도 :0.688, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\sklearn\utils\validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


4 교차검증 정확도 :0.6787, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\sklearn\utils\validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


5 교차검증 정확도 :0.674, 학습 데이터 크기:12000, 검증 데이터 크기:3000

 평균검증 정확도: 0.68186


In [164]:
# 결정트리를 사용하여 정확도 확인

from sklearn.tree import DecisionTreeClassifier

dtree=DecisionTreeClassifier(random_state=0)

In [165]:
### 교차검증

from sklearn.model_selection import KFold
from sklearn.metrics import accuracy_score

kfold=KFold(n_splits=5)
n_iter=0
r_accuracy=[]

for train_index, test_index in kfold.split(r_feature):
    X_train, X_test=r_feature[train_index],r_feature[test_index]
    Y_train, Y_test=r_target[train_index],r_target[test_index]
    
    dtree.fit(X_train,Y_train)
    pred=dtree.predict(X_test)
    n_iter+=1
    
    accuracy=np.round(accuracy_score(Y_test,pred),4)
    train_size=X_train.shape[0]
    test_size=X_test.shape[0]
    print('{0} 교차검증 정확도 :{1}, 학습 데이터 크기:{2}, 검증 데이터 크기:{3}'.format(n_iter,accuracy,train_size,test_size))
    
    r_accuracy.append(accuracy)
    
print('\n 평균검증 정확도:',np.mean(r_accuracy))

1 교차검증 정확도 :0.7993, 학습 데이터 크기:12000, 검증 데이터 크기:3000
2 교차검증 정확도 :0.8067, 학습 데이터 크기:12000, 검증 데이터 크기:3000
3 교차검증 정확도 :0.809, 학습 데이터 크기:12000, 검증 데이터 크기:3000
4 교차검증 정확도 :0.7893, 학습 데이터 크기:12000, 검증 데이터 크기:3000
5 교차검증 정확도 :0.7937, 학습 데이터 크기:12000, 검증 데이터 크기:3000

 평균검증 정확도: 0.7996


In [170]:
# KNN을 사용하여

from sklearn.neighbors import KNeighborsClassifier

best_k=[0,0]
for i in range(2,11):
    knn = KNeighborsClassifier(n_neighbors=i)
    
    
    kfold=KFold(n_splits=5)
    n_iter=0
    r_accuracy=[]

    for train_index, test_index in kfold.split(r_feature):
        X_train, X_test=r_feature[train_index],r_feature[test_index]
        Y_train, Y_test=r_target[train_index],r_target[test_index]
    
        knn.fit(X_train,Y_train)
        pred=knn.predict(X_test)
        n_iter+=1
    
        accuracy=np.round(accuracy_score(Y_test,pred),4)
        train_size=X_train.shape[0]
        test_size=X_test.shape[0]
        print('{0} 교차검증 정확도 :{1}, 학습 데이터 크기:{2}, 검증 데이터 크기:{3}'.format(n_iter,accuracy,train_size,test_size))
    
        r_accuracy.append(accuracy)
        
        if best_k[0]<np.mean(r_accuracy):
            best_k[0]=np.mean(r_accuracy)
            best_k[1]=i

print('k={0} \n 평균검증 정확도:{1}'.format(best_k[1],best_k[0]))

C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


1 교차검증 정확도 :0.7637, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


2 교차검증 정확도 :0.7597, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


3 교차검증 정확도 :0.7687, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


4 교차검증 정확도 :0.7557, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


5 교차검증 정확도 :0.7563, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


1 교차검증 정확도 :0.777, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


2 교차검증 정확도 :0.773, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


3 교차검증 정확도 :0.7757, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


4 교차검증 정확도 :0.7743, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


5 교차검증 정확도 :0.7623, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


1 교차검증 정확도 :0.7837, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


2 교차검증 정확도 :0.777, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


3 교차검증 정확도 :0.7823, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


4 교차검증 정확도 :0.7793, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


5 교차검증 정확도 :0.773, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


1 교차검증 정확도 :0.7793, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


2 교차검증 정확도 :0.7783, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


3 교차검증 정확도 :0.783, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


4 교차검증 정확도 :0.7783, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


5 교차검증 정확도 :0.7773, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


1 교차검증 정확도 :0.7877, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


2 교차검증 정확도 :0.7877, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


3 교차검증 정확도 :0.789, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


4 교차검증 정확도 :0.783, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


5 교차검증 정확도 :0.7753, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


1 교차검증 정확도 :0.7817, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


2 교차검증 정확도 :0.7873, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


3 교차검증 정확도 :0.7883, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


4 교차검증 정확도 :0.783, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


5 교차검증 정확도 :0.7707, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


1 교차검증 정확도 :0.7807, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


2 교차검증 정확도 :0.7837, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


3 교차검증 정확도 :0.7917, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


4 교차검증 정확도 :0.7817, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


5 교차검증 정확도 :0.7737, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


1 교차검증 정확도 :0.775, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


2 교차검증 정확도 :0.788, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


3 교차검증 정확도 :0.793, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


4 교차검증 정확도 :0.7823, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


5 교차검증 정확도 :0.776, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


1 교차검증 정확도 :0.7783, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


2 교차검증 정확도 :0.789, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


3 교차검증 정확도 :0.7917, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


4 교차검증 정확도 :0.7837, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:18: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().


5 교차검증 정확도 :0.7753, 학습 데이터 크기:12000, 검증 데이터 크기:3000
k=6 
 평균검증 정확도:0.7881333333333332


In [171]:
from sklearn.naive_bayes import GaussianNB

nb = GaussianNB()

In [172]:
### 교차검증

from sklearn.model_selection import KFold
from sklearn.metrics import accuracy_score

kfold=KFold(n_splits=5)
n_iter=0
r_accuracy=[]

for train_index, test_index in kfold.split(r_feature):
    X_train, X_test=r_feature[train_index],r_feature[test_index]
    Y_train, Y_test=r_target[train_index],r_target[test_index]
    
    nb.fit(X_train,Y_train)
    pred=nb.predict(X_test)
    n_iter+=1
    
    accuracy=np.round(accuracy_score(Y_test,pred),4)
    train_size=X_train.shape[0]
    test_size=X_test.shape[0]
    print('{0} 교차검증 정확도 :{1}, 학습 데이터 크기:{2}, 검증 데이터 크기:{3}'.format(n_iter,accuracy,train_size,test_size))
    
    r_accuracy.append(accuracy)
    
print('\n 평균검증 정확도:',np.mean(r_accuracy))

1 교차검증 정확도 :0.6817, 학습 데이터 크기:12000, 검증 데이터 크기:3000
2 교차검증 정확도 :0.6777, 학습 데이터 크기:12000, 검증 데이터 크기:3000
3 교차검증 정확도 :0.687, 학습 데이터 크기:12000, 검증 데이터 크기:3000
4 교차검증 정확도 :0.674, 학습 데이터 크기:12000, 검증 데이터 크기:3000
5 교차검증 정확도 :0.6653, 학습 데이터 크기:12000, 검증 데이터 크기:3000

 평균검증 정확도: 0.67714


C:\Anaconda3\lib\site-packages\sklearn\utils\validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
C:\Anaconda3\lib\site-packages\sklearn\utils\validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
C:\Anaconda3\lib\site-packages\sklearn\utils\validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
C:\Anaconda3\lib\site-packages\sklearn\utils\validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)
C:\Anaconda3\lib\site-packages\sklearn\u

In [198]:
from sklearn.ensemble import StackingClassifier
import lightgbm


def get_stacking():
    level0=[]
    level0.append(('svm',svm))
    level0.append(('knn',KNeighborsClassifier(n_neighbors=6)))
    level0.append(('dtree',dtree))
    
    level1=lightgbm.LGBMClassifier(n_estimators=400,learning_rate=0.1,max_depth=3)
    
    model=StackingClassifier(estimators=level0, final_estimator=level1,cv=3)
    
    return model

stack=get_stacking()

In [206]:
### 교차검증

from sklearn.model_selection import KFold
from sklearn.metrics import accuracy_score

kfold=KFold(n_splits=5)
n_iter=0
r_accuracy=[]

for train_index, test_index in kfold.split(r_feature):
    X_train, X_test=r_feature[train_index],r_feature[test_index]
    Y_train, Y_test=r_target[train_index],r_target[test_index]
    
    stack.fit(X_train,Y_train)
    pred=dtree.predict(X_test)
    n_iter+=1
    
    accuracy=np.round(accuracy_score(Y_test,pred),4)
    train_size=X_train.shape[0]
    test_size=X_test.shape[0]
    print('{0} 교차검증 정확도 :{1}, 학습 데이터 크기:{2}, 검증 데이터 크기:{3}'.format(n_iter,accuracy,train_size,test_size))
    
    r_accuracy.append(accuracy)
    
print('\n 평균검증 정확도:',np.mean(r_accuracy))

C:\Anaconda3\lib\site-packages\sklearn\utils\validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


1 교차검증 정확도 :0.9993, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\sklearn\utils\validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


2 교차검증 정확도 :0.9983, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\sklearn\utils\validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


3 교차검증 정확도 :0.9993, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\sklearn\utils\validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


4 교차검증 정확도 :0.9987, 학습 데이터 크기:12000, 검증 데이터 크기:3000


C:\Anaconda3\lib\site-packages\sklearn\utils\validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


5 교차검증 정확도 :0.7937, 학습 데이터 크기:12000, 검증 데이터 크기:3000

 평균검증 정확도: 0.9578599999999999


# 5. 테스트 Case 결과 확인

In [175]:
f_test=pd.read_csv(PATH+'test_err_data.csv')

In [176]:
f_test.head(5)

user_id            time model_nm       fwver  errtype errcode
0    30000  20201101030227  model_1  04.16.3553       31       1
1    30000  20201101030227  model_1  04.16.3553       33       2
2    30000  20201101030228  model_1  04.16.3553       15       1
3    30000  20201101030256  model_1  04.16.3553       22       1
4    30000  20201101030300  model_1  04.16.3553       11       1

In [177]:
f_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16532648 entries, 0 to 16532647
Data columns (total 6 columns):
 #   Column    Dtype 
---  ------    ----- 
 0   user_id   int64 
 1   time      int64 
 2   model_nm  object
 3   fwver     object
 4   errtype   int64 
 5   errcode   object
dtypes: int64(3), object(3)
memory usage: 756.8+ MB


In [188]:
len(f_test.user_id.unique())

14998

In [179]:
f_test.errtype.unique()

array([31, 33, 15, 22, 11, 12, 23, 26, 16, 17,  5, 39, 24, 40, 36, 37, 32,
       42, 14, 18, 25, 20, 35,  6,  7,  4,  3, 13, 27, 28, 41, 34, 38,  1,
        8, 21, 19,  2, 10, 30,  9], dtype=int64)

In [180]:
test_errtype_feature=[]
test_tmp=[0]*41
        
for i in range(1,len(f_test)):
    if f_test['user_id'].values[i]==f_test['user_id'].values[i-1]:
        if f_test.errtype.values[i]<29:
            test_tmp[f_test.errtype.values[i]-1]+=1
        else:
            test_tmp[f_test.errtype.values[i]-2]+=1

    else:
        test_errtype_feature.append(test_tmp)
        test_tmp=[0]*41
        if f_test.errtype.values[i]<29:
            test_tmp[f_test.errtype.values[i]-1]+=1
        else:
            test_tmp[f_test.errtype.values[i]-2]+=1

test_errtype_feature.append(test_tmp)

In [189]:
r_test_feature=np.array(test_errtype_feature)
r_test_feature.shape

(14998, 41)